internal/datastore/proxy/relationshipintegrity.go

```diff
@@ -0,0 +1,384 @@
+package proxy
+
+import (
+	"context"
+	"crypto/hmac"
+	"fmt"
+	"hash"
+	"time"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/authzed/spicedb/pkg/datastore"
+	"github.com/authzed/spicedb/pkg/datastore/options"
+	corev1 "github.com/authzed/spicedb/pkg/proto/core/v1"
+	"github.com/authzed/spicedb/pkg/spiceerrors"
+	"github.com/authzed/spicedb/pkg/tuple"
+	"github.com/minio/highwayhash"
+)
+
+// KeyConfig is a configuration for a key used to sign relationships.
+type KeyConfig struct {
+	ID        string
+	ExpiredAt *time.Time
+	Bytes     []byte
+}
+
+type hmacConfig struct {
+	keyID     string
+	expiredAt *time.Time
+	hasher    hash.Hash
+}
+
+var (
+	versionByte = byte(0x01)
+)
+
+// NewRelationshipIntegrityProxy creates a new datastore proxy that ensures the integrity of
+// relationships by using HMACs to sign the data. The current key is used to sign new data,
+// and the expired keys are used to verify old data, if any.
+func NewRelationshipIntegrityProxy(ds datastore.Datastore, currentKey KeyConfig, expiredKeys []KeyConfig) (datastore.Datastore, error) {
+	hasher, err := hasherForKey(currentKey.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	currentKeyHMAC := hmacConfig{
+		keyID:     currentKey.ID,
+		expiredAt: currentKey.ExpiredAt,
+		hasher:    hasher,
+	}
+
+	if currentKey.ExpiredAt != nil {
+		return nil, spiceerrors.MustBugf("current key cannot have an expiration")
+	}
+
+	keysByID := make(map[string]hmacConfig, len(expiredKeys)+1)
+	keysByID[currentKey.ID] = currentKeyHMAC
+
+	for _, key := range expiredKeys {
+		if key.ExpiredAt == nil {
+			return nil, spiceerrors.MustBugf("expired key missing expiration time")
+		}
+
+		if _, ok := keysByID[key.ID]; ok {
+			return nil, spiceerrors.MustBugf("found duplicate key ID: %s", key.ID)
+		}
+
+		hasher, err := hasherForKey(key.Bytes)
+		if err != nil {
+			return nil, err
+		}
+
+		keysByID[key.ID] = hmacConfig{
+			keyID:     key.ID,
+			expiredAt: key.ExpiredAt,
+			hasher:    hasher,
+		}
+	}
+
+	return &relationshipIntegrityProxy{
+		ds:         ds,
+		primaryKey: currentKeyHMAC,
+		keysByID:   keysByID,
+	}, nil
+}
+
+func hasherForKey(key []byte) (hash.Hash, error) {
+	return highwayhash.New64(key)
+}
+
+type relationshipIntegrityProxy struct {
+	ds         datastore.Datastore
+	primaryKey hmacConfig
+	keysByID   map[string]hmacConfig
```

Comment:
it's usually a good idea to encode the hash algorithm/parameters in the key id for later retrieval (password hashing usually does this for example) so that the algorithm / parameters can be rotated alongside the keys themselves.

I.e. it's sha256 now but we decide we need to rotate to sha512 or sha3 etc.

Reply:
That's why we have the algorithm version byte; we can map from that to the alg used

Reply:
Missed that, sorry